### PR TITLE
Implements TkProduces decorator as for #108 

### PR DESCRIPTION
**Pull request by ekondrashev**
Referenced to #108 issue.
Please take a look at this PR implementing `TkProduces` only since `TsProduces` is no longer relevant.

I've implemented this making use of `RsFork` and `FkTypes`.

Please take a look at `throwsExceptionOnUnsupportedAcceptHeader` test - I'm not sure that it is good idea to not fail fast in case no proper accept header handler found.

### CODE REVIEW

**Comment by davvd** [comment]
Thanks, I will find someone to review this PR soon

**Comment by davvd** [comment]
@pinaf this pull request is for you, please review

**Comment by pinaf** [comment]
@ekondrashev please format your ticket according to the QA policy http://www.teamed.io/qa.html
- please make a reference to the underlying issue this is fixing
- **Pull request title explains the problem it is fixing**

File: src/main/java/org/takes/facets/fork/TkProduces.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.takes.tk.TkWrap;
+
+/**
+ * Take that acts on request with specified "Accept" HTTP headers only.
+ *
+ * <p>The class is immutable and thread-safe.
+ *
+ * @author Eugene Kondrashev (eugene.kondrashev@gmail.com)
+ * @version $Id$
+ * @since 0.32.1
```
Comment (pinaf) [inline src/main/java/org/takes/facets/fork/TkProduces.java]: @ekondrashev this version is not right. 0.14 maybe?

File: src/test/java/org/takes/facets/fork/TkProducesTest.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import org.takes.rs.RsJSON;
+import org.takes.rs.RsPrint;
+import org.takes.tk.TkEmpty;
+import org.takes.tk.TkFixed;
+
+/**
+ * Test case for {@link TkProduces}.
+ * @author Eugene Kondrashev (eugene.kondrashev@gmail.com)
+ * @version $Id$
+ * @since 0.31.1
```
Comment (pinaf) [inline src/test/java/org/takes/facets/fork/TkProducesTest.java]: @ekondrashev same comment about version

File: src/test/java/org/takes/facets/fork/TkProducesTest.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ */
+public final class TkProducesTest {
+
+    /**
+     * TkProduces can throw exception on unsupported Accept header.
+     * @throws IOException If some problem inside
+     */
+    @Test(expected = HttpException.class)
+    public void throwsExceptionOnUnsupportedAcceptHeader() throws IOException {
```
Comment (pinaf) [inline src/test/java/org/takes/facets/fork/TkProducesTest.java]: @ekondrashev let's call this `failsOnUnsupportedAcceptHeader`.

File: src/test/java/org/takes/facets/fork/TkProducesTest.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+/**
+ * Test case for {@link TkProduces}.
+ * @author Eugene Kondrashev (eugene.kondrashev@gmail.com)
+ * @version $Id$
+ * @since 0.31.1
+ */
+public final class TkProducesTest {
+
+    /**
+     * TkProduces can throw exception on unsupported Accept header.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/fork/TkProducesTest.java]: @ekondrashev let's go with **TkProduces can fail on unsupported Accept header.**

**Comment by pinaf** [comment]
@ekondrashev looks very good! 4 minor comments above. Also, I don't undertand your concern about the test - it looks like it fails fast by throwing the exception.

**Comment by pinaf** [comment]
@ekondrashev thanks. Are you done?

**Comment by ekondrashev** [comment]
@pinaf yes, I'm done, thank you.

**Comment by ekondrashev** [comment]
@pinaf as for not failing fast concern - I meant that we should not create response if it is not suitable for target request. Currently it is failing on Response.head() call, but I can't see how it can be implemented differently considering RsFork implementation.

**Comment by pinaf** [comment]
@rultor merge

File: src/test/java/org/takes/facets/fork/TkProducesTest.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+ */
+public final class TkProducesTest {
+
+    /**
+     * TkProduces can fail on unsupported Accept header.
+     * @throws IOException If some problem inside
+     */
+    @Test(expected = HttpException.class)
+    public void failsOnUnsupportedAcceptHeader() throws IOException {
+        final TkProduces produces = new TkProduces(
```
Comment (yegor256) [inline src/test/java/org/takes/facets/fork/TkProducesTest.java]: it's enough to do `final Take produces`. your IDE should give you a warning here... try to use these settings for IntelliJ: http://img.teamed.io/settings.jar

File: src/test/java/org/takes/facets/fork/TkProducesTest.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        ).head();
+    }
+
+    /**
+     * TkProduce can produce correct type response.
+     * @throws IOException If some problem inside
+     */
+    @Test
+    public void producesCorrectContentTypeResponse() throws IOException {
+        final TkProduces produces = new TkProduces(
```
Comment (yegor256) [inline src/test/java/org/takes/facets/fork/TkProducesTest.java]: same here

**Comment by yegor256** [comment]
@ekondrashev a few comments from me

**Comment by ekondrashev** [comment]
@yegor256 done with your comments, thank you

**Comment by yegor256** [comment]
@rultor merge

**Comment by pinaf** [comment]
@yegor256 looks unrelated. try again?

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@elenavolokhova please, let us know what do you think about this ticket, according to our [QA rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor please deploy

**Comment by elenavolokhova** [comment]
@davvd This ticket looks good! :)

**Comment by davvd** [comment]
> @davvd This ticket looks good! :)

@elenavolokhova thank you, good quality is always good :)

**Comment by davvd** [comment]
@pinaf I added 10 mins to @elenavolokhova (for QA review) in transaction 55999403... thank you, added **20 mins** to your acc, payment num is `AP-51H584401R045221Y`, 213 hours spent in total to complete... review comments (c=5) added as a bonus... added +20 to your rating, now it is equal to [+7815](http://www.netbout.com/b/35128?open=rating)

**Pull request merged**